### ISSUE TITLE
Partitions in AWS Athena 

### ISSUE THREAD

**Issue by sh-rp** [dlt-hub/dlt#555]
We want to enable the user to make use of AWS Athena partitions. For this we will most likely need to change the schema a bit and add the option for the user to define partitions on the resources. 

Tasks:
- [ ] Alter schema to have an order of partitions (possibly replace the "bool" hint with an "int" hint which is null or 0 if this column should not be a partition, and a number if it should which then defines the order?)
- [ ] Expose partition config on the resource as array of strings which are the column names?
- [ ] When creating parquet files, respect the partitions and create a file for every partition key combination and upload that to the right path.
- [ ] Make the partitions known to Athena, either we can run explicit ddl commands, but it should also be possible to infer the partitions with a repair command, the latter is more costly though, so doing this with our schema makes sense. This means we will have schema m migration operations that not only add columns but also partition definitions.

To think about:
- [ ] How to guard against the user uploading to the same bucket with different partition settings?
- [ ] How to migrate existing schemas if we change the type of the partition settings?
- [ ] How to behave if there are partitions defined, but the yielded resource item does not have a value for this column? Will there always be a "None" path for those cases? Will this be queryable by Athena?

**Comment by gamgi** [comment]
Documenting a potential use-case. I'm loading files from S3 + SFTP sources to S3 destination.

I'd like to use properties from the source (not data itself) to design the partition layout in the destination. The properties may be resource *arguments*, or indirectly source metadata like  *object path* and/or *file name*.

Examples:
* Single daily object from SFTP
  * `sftp://example.com/foo/bar/report_20231114000000.csv` --> `s3://destination-bucket/baz/quz/yyyy=2023/mm=11/dd=14/report.csv`
* Multiple daily objects from S3
  * `s3://example-bucket/foo/bar/report_20231114000000.csv` --> `s3://destination-bucket/baz/quz/yyyy=2023/mm=11/dd=14/report_20231114000000.csv`
  * `s3://example-bucket/foo/bar/report_20231114005000.csv` --> `s3://destination-bucket/baz/quz/yyyy=2023/mm=11/dd=14/report_20231114005000.csv`
* Single daily API call to S3
  * `https://example.com/api/report?date=2023-11-14` --> `s3://destination-bucket/baz/quz/yyyy=2023/mm=11/dd=14/report.csv`


Implications for this ticket:
* Partition config should support resource parameters
    * For example, if `date` (or `year, month, date`) is a parameter to a resource, that should be available in the partition layout.
* Partition config should support resource metadata, such as *object path* or *creation date*.
  * For example, in order to replicate the partitioning from the source to destination.
* Consider: Is partitioning a generic issue, or something purely related to filesystem destination?

TL;DR I'm looking to partition based on metadata, not the data itself.

**Comment by MRocholl** [comment]
Just wanted to add a thing to think about.

Iceberg tables make use of hidden partitioning. And keep track of different partitioning over time. 
This might be much simpler to implement than the classical hive table. 

https://trino.io/blog/2021/05/03/a-gentle-introduction-to-iceberg#hidden-partitions

Considering this issue is about AWS Athena partitions I do not believe this is out of scope and might be the "simpler" part of the issue.

Best

**Comment by Pipboyguy** [comment]
Subsumed by https://github.com/dlt-hub/dlt/issues/930

**Comment by sultaniman** [comment]
@nicor88 I've been working on [extending layout placeholders](https://github.com/dlt-hub/dlt/pull/1182) for filesystem destination, atm we don't yet support syncing or taking date & time from parquet files for partitioning.

**Comment by nicor88** [comment]
Layout placehoders for what I can see won't work with Iceberg athena managed tables. 

For example an iceberg table can be created like that:
```
CREATE TABLE events (id bigint, event_date date)
  PARTITIONED BY (event_date, bucket(16, id))
  LOCATION 's3://DOC-EXAMPLE-BUCKET/your-folder/'
  TBLPROPERTIES ( 'table_type' = 'ICEBERG' )
```
The writing to the file systems is handle under the hood by athena itself, not by the writer, this because of Iceberg hidden partitioning.
Therefore the only thing to do with athena iceberg tables is to specify the partition definition on table creation.

Said so, I was planning to propose a PR for that, because it's the only limitation that is blocking us to use dlt in production.

All our tables are partitioned by ingestion date for example, to reduce data scans down-streams.

**Comment by sultaniman** [comment]
@nicor88 sure we can have a chat about it, I think once we merge the original PR I will be more that happy for allowing iceberg support 🙂

**Comment by rudolfix** [comment]
athena partitions got implemented in https://github.com/dlt-hub/dlt/pull/1403
